로그인기능에 rx 적용, 텍스트필드 감시해서 로그인버튼 활성화 바인딩

torpedo87 opened the pull request.

Commit:
Userdefaults에 토큰 저장시 Completable 적용

Commit:
Make login process reactive
func getToken 이 Status 반환하기 fix #2
id, password 텍스트필드 감시해서 로그인 버튼 활성화 fix #3

hoemoon commented on uIssue/Controller/LoginViewController.swift:
```diff
@@ -60,21 +62,47 @@ class LoginViewController: UIViewController {
+        return true
+    }
+    
+    let validate = Observable.combineLatest(idInput, pwdInput)
```
validate 따로 왜 만든거지?

<sub>_commented via [Diff Monster](https://diff.sapzil.org/#/torpedo87/uIssue/pull/4?path=uIssue%2FController%2FLoginViewController.swift)_</sub>

torpedo87 replied:
수정했어요

hoemoon commented on uIssue/Service/UserNetworkManager.swift:
```diff
@@ -14,9 +14,25 @@ enum Errors: Error {
   case invalidUserInfo
 }
 
+enum Status: Equatable {
```
매니저 내부에 enum 만들면 인터페이스가 더 깔끔해져요

<sub>_commented via [Diff Monster](https://diff.sapzil.org/#/torpedo87/uIssue/pull/4?path=uIssue%2FService%2FUserNetworkManager.swift)_</sub>

torpedo87 replied:
받들겠습니다

hoemoon commented on uIssue/Extensions/Userdefaults.swift:
```diff
@@ -7,12 +7,25 @@
-  func saveToken(token: Token) {
-    let dict = token.asDictionary
-    UserDefaults.standard.set(dict, forKey: "token")
+  func saveToken(token: Token) -> Completable {
```
Token을 UserDefaults로 바로 접근하는 것보다 status 같은 enum에 연관값?으로 넣어서 다루는 게 더 깔끔할 것 같아요

<sub>_commented via [Diff Monster](https://diff.sapzil.org/#/torpedo87/uIssue/pull/4?path=uIssue%2FExtensions%2FUserdefaults.swift)_</sub>

torpedo87 replied:
무슨 뜻인지 잘 모르겠어요

hoemoon replied:
saveToken에서 Completable 반환하는 게 좀 이상한 거 같고, 매니저 같은데서만 공개 인터페이스를 갖는 게 좋아 보인다는 이야기에요 ㅋㅋ

hoemoon commented on uIssue/Controller/LoginViewController.swift:
```diff
@@ -60,21 +62,47 @@ class LoginViewController: UIViewController {
+      .flatMap { [weak self] _ -> Single<Status> in
         UserNetworkManager
-            .getToken(userId: self!.idTextField.text!, userPassword: self!.passWordTextField.text!)
+          .getToken(userId: (self?.idTextField.text)!, userPassword: (self?.passWordTextField.text)!)
```
마찬 가지로 status 만든 거면 enum 에 넣어서 반환하는 게 더 나을 듯.

<sub>_commented via [Diff Monster](https://diff.sapzil.org/#/torpedo87/uIssue/pull/4?path=uIssue%2FController%2FLoginViewController.swift)_</sub>

torpedo87 replied:
무슨 뜻인지 잘 모르겠어요

hoemoon commented on uIssue/Controller/LoginViewController.swift:
```diff
@@ -60,21 +62,47 @@ class LoginViewController: UIViewController {
-      .asDriver(onErrorJustReturn: Token(id: -1, token: "error"))
-      .drive(onNext: { [weak self] token in
-        if token.isValid() {
-          UserDefaults.standard.saveToken(token: token)
```
이 레벨에서 UserDefault다루는 게 이상해요 감추는 게 좋은 듯

<sub>_commented via [Diff Monster](https://diff.sapzil.org/#/torpedo87/uIssue/pull/4?path=uIssue%2FController%2FLoginViewController.swift)_</sub>

torpedo87 replied:
이부분은 삭제한 부분입니당

Commit:
Pullreq 수정
enum 타입은 클래스 안에 넣기
로그인 요청시 request를 Observable 로 감시하기
RepoListVC에 설정버튼 달기
validate 참조 없애기
'
